src/main/kotlin/org/rust/lang/utils/RustEscapesUtils.kt

```diff
@@ -0,0 +1,89 @@
+
+import com.intellij.psi.StringEscapesTokenTypes.*
+import org.rust.lang.core.lexer.RustEscapesLexer
+import org.rust.lang.core.psi.RustTokenElementTypes.STRING_LITERAL
+
+/**
+ * Unescape string escaped using Rust escaping rules.
+ */
+fun String.unescapeRust(unicode: Boolean = true, eol: Boolean = true): String {
+    val sb = StringBuilder(length)
+    val lexer = RustEscapesLexer(STRING_LITERAL, unicode, eol)
+    lexer.start(this)
+    while (lexer.tokenType != null) {
+        sb.append(when (lexer.tokenType) {
+            VALID_STRING_ESCAPE_TOKEN -> decodeEscape(lexer.tokenText)
+            else                      -> lexer.tokenText
+        })
+        lexer.advance()
+    }
+    return sb.toString()
+}
+
+/**
+ * Mimics [com.intellij.codeInsight.CodeInsightUtilCore.parseStringCharacters], but obeys Rust escaping rules.
+ */
+fun parseRustStringCharacters(chars: String, outChars: StringBuilder, sourceOffsets: IntArray? = null): Boolean {
+    assert(sourceOffsets == null || sourceOffsets.size == chars.length + 1)
+
+    val lexer = RustEscapesLexer(STRING_LITERAL, true, true)
+    lexer.start(chars)
+
+    val outOffset = outChars.length
+    var index = 0
+    while (lexer.tokenType != null) {
+        when (lexer.tokenType) {
+            VALID_STRING_ESCAPE_TOKEN    -> {
+                outChars.append(decodeEscape(lexer.tokenText))
+                if (sourceOffsets != null) {
+                    // Set offset for the decoded character to the beginning of the escape sequence.
+                    sourceOffsets[outChars.length - outOffset - 1] = index
+                    // And perform a "jump"
+                    index += lexer.tokenEnd - lexer.tokenStart
+                }
+            }
+
+            INVALID_CHARACTER_ESCAPE_TOKEN,
+            INVALID_UNICODE_ESCAPE_TOKEN ->
+                return false
+
+            else                         -> {
+                val first = outChars.length - outOffset
+                outChars.append(lexer.tokenText)
+                val last = outChars.length - outOffset - 1
+                if (sourceOffsets != null) {
+                    // Set offsets for each character of given chunk
+                    for (i in first..last) {
+                        sourceOffsets[i] = index
+                        index++
+                    }
+                }
+            }
+        }
+        lexer.advance()
+    }
+
+    return true
+}
+
+private fun decodeEscape(esc: String): String = when (esc) {
+    "\\n"  -> "\n"
+    "\\r"  -> "\r"
+    "\\t"  -> "\t"
+    "\\0"  -> "\u0000"
+    "\\\\" -> "\\"
+    "\\'"  -> "\'"
+    "\\\"" -> "\""
+
+    else   -> {
+        assert(esc.length >= 2)
+        assert(esc[0] == '\\')
+        when (esc[1]) {
+            'x'        -> Integer.parseUnsignedInt(esc.substring(2), 16).toChar().toString()
+            'u'        -> Integer.parseUnsignedInt(esc.substring(3, esc.length - 1), 16).toChar().toString()
+            '\r', '\n' -> ""
```

Comment:
Ok, one more concern: I remember that escapes like `"\n         "` (new line and several spaces, like [here](https://github.com/intellij-rust/intellij-rust/pull/249/files#diff-31373ef0f4bfe0155469f9c0dfe5a8dfR19)) are possible. Should this be handled here?